### ISSUE TITLE
Replace prompt embedding methods with `encode_prompt()`

### ISSUE THREAD

**Issue by sayakpaul** [a-r-r-o-w/finetrainers#221]
Currently, we make use of adapted / copy-pasted utilities from `Diffusers` to embed captions. But since `encode_prompt()` method of a pipeline is flexible, we could use that directly. It will help us reduce LoC quite a bit and have a nice separation of concerns (in this case we delegate `encode_prompt()` to `diffusers`). 

@a-r-r-o-w would you be open to supporting this? If so, I can make a quick PoC PR to get feedback and ship.

**Comment by a-r-r-o-w** [comment]
Thanks for the suggestion. The design we have currently is carefully thought of and part of broader design goal and so I don't want to rely on using diffusers `encode_prompt` at the moment, or in future, because:
- they contain additional logic that is not relevant to training
- we cannot really control or modify the encode prompt logic for custom text preprocessing pipeline
- feel free to disagree but it is easier to read the code of entire training related utilities in a single file like we have currently imo

As an example of what I envision custom text preprocessing pipeline to be, think about how a ComfyUI workflow works. You can have various nodes that handle preprocessing before the underlying things you're working with is queued for actual processing. Each part is modular and performs exactly one thing, and these different parts can be combined in various ways to perform preprocessing as you like, be modified or opted out of on-the-fly, etc. This is to say that there are caption filtering/dropout/modification/etc. methods proposed in various papers and I would like to implement them in a modular manner, with users being able to opt-in to them when needed (or choosing different strategy for different dataset being trained on - style/concept/character lora training all require different kinds of caption preprocessing for good training results). The `encode_prompt` method, while being great for just going from text to embeddings, will divert the code flow out of finetrainers - this is something I want to limit as much as possible so that more things are under our control. I don't want to tie the design too closely with diffusers for two reasons:
- In its current implementation, it might be hard to see but the broader goal for me is to be able to provide an API where anyone can hook in a custom model and utilities (as some specification we design) and train it with all the goodies we offer: loras, full finetuning/training, distillation, distributedness with DDP/FSDP/PP/TP/CP blackboxed, etc.
- The lesser we tie in upstream dependencies, the lesser prone we are to breaking changes. Also, having all code to read at one place is good for educational purposes and I find it really helpful. For this reason and as an example, any custom triton kernels I introduce for memory/speed efficiency eventually, will either be implemented/benchmarked by ourselves or be copied verbatim from upstream library (for example, sage attention). I do not see this as a burden of maintainence because we have more control over modifying things to fit our design and work with all models we want to support OOTB

**Comment by sayakpaul** [comment]
All good, thanks a lot for explaining!
